### ISSUE TITLE
Limited converage of 3d conv. with no-workspace-required kernels.

### ISSUE THREAD

**Issue by zjing14** [ROCmSoftwarePlatform/MIOpen#588]
For SWDEV-258292, we cannot rely on GEMM, which requires large workspaces, as fallbacks for 3d conv.

For the 3DUNet medical benchmark in the JIRA, we miss the support of the following 3d conv. configs with no-GEMM kernels.

`./bin/test_conv3d --disable-validation --float --cmode conv --pmode default --group-count 1 --input 2, 32, 128, 128, 128 --weights 4, 32, 3, 3, 3 --pads_strides_dilations 1 1 1 1 1 1 1 1 1 --disable-backward-data --disable-backward-weights` (fwd, wrw)
`./bin/test_conv3d --float --cmode conv --pmode default --group-count 1 --input 2, 4, 128, 128, 128 --weights 32, 4, 3, 3, 3 --pads_strides_dilations 1 1 1 1 1 1 1 1 1 --trans_output_pads 0 0 0` (bwd, wrw)

**Comment by atamazov** [comment]
A naive implementation is now being added to the driver (#583). I think we can use it for 3D right away, right?

**Comment by carlushuang** [comment]
@atamazov Hi that can be used, but that is very un-optimized, and don't expect good performance. I think maybe we still need a implicit gemm implementation

**Comment by zjing14** [comment]
@atamazov @carlushuang Let me try. It will be great if we have some algorithms to support it before we have igemm implementation.

**Comment by zjing14** [comment]
@carlushuang I tried with the reference kernel, still not find a suitable algorithm. Did I use the reference kernel correctly?
`MIOPEN_FIND_MODE=1 MIOPEN_DRIVER_USE_GPU_REFERENCE=1 MIOPEN_DEBUG_CONV_GEMM=0 ./bin/test_conv3d --float --cmode conv --pmode default --group-count 1 --input 2, 32, 128, 128, 128 --weights 4, 32, 3, 3, 3 --pads_strides_dilations 1 1 1 1 1 1 1 1 1
FAILED: /dockerx/MIOpen/src/ocl/convolutionocl.cpp:791: Forward Convolution cannot be executed due to incorrect params`

**Comment by carlushuang** [comment]
@zjing14 Yes, currently we just use it for validation. For if it can be a fallback kernel or not, the answer is yes. But the only concern is performance. Because you can check [naive_conv.cpp](https://github.com/ROCmSoftwarePlatform/MIOpen/blob/e91e4e8f07a279c8571829943af1cb8acb51df9c/src/kernels/gpu_reference_kernel/naive_conv.cpp), there is no optimization at all.

**Comment by zjing14** [comment]
@carlushuang Yes. We definitely will need the support of igemm. But currently, we may need reference kernels as fallback to resolve the issue as soon as possible. Could you add reference kernels into MIOpen as an algorithm?

**Comment by carlushuang** [comment]
@zjing14 OK, let me think about this feature in PR: #583

**Comment by zjing14** [comment]
@carlushuang At this moment, We may simply add it as a solver in to direct conv. Algorithm rather than a fallback as GEMM. It should be simpler than replacing GEMM as a fallback. @atamazov What do you think?

**Comment by carlushuang** [comment]
Yes... and if have an algorithm name like `NAIVE_CONV` is very strange.

**Comment by atamazov** [comment]
Of course, this is another Solver that implements Direct algo. Just add it to begin with.

Then open a ticket where we can discuss and decide how to integrate it into Immediate Mode Fallback. When decision is made, the implementation will not be a problem.

**Comment by carlushuang** [comment]
@zjing14 @atamazov  this reference kernel is added as a solver together in PR:https://github.com/ROCmSoftwarePlatform/MIOpen/pull/583, and open an issue:https://github.com/ROCmSoftwarePlatform/MIOpen/issues/593 for further discussion

**Comment by ppanchad-amd** [comment]
@zjing14 Please check if this still an issue on ROCm 6.0.2? If not reproducible, please close ticket. Thanks!

**Comment by atamazov** [comment]
@ppanchad-amd This is implemented and may be closed.
